### ISSUE TITLE
Error compiling kernel modules under arm32 bit

### ISSUE THREAD

**Issue by ewildgoose** [vel21ripn/nDPI#192]
Hi, compiling on arm32 I immediately see an error:
```
make  -C /usr/src/linux M=$PWD modules -j32 --jobserver-auth=fifo:/tmp/GMfifo2018;
make[2]: Entering directory '/usr/src/linux'
make[2]: warning: -j32 forced in submake: resetting jobserver mode.
In file included from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h:329,
                 from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_typedefs.h:44,
                 from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/lib/third_party/include/ndpi_sha256.h:16,
                 from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/lib/third_party/src/ndpi_sha256.c:27:
/usr/lib/gcc/armv7a-unknown-linux-musleabihf/13/include/stdint.h:9:26: error: no include path in which to search for stdint.h
    9 | # include_next <stdint.h>
      |                          ^
/var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h: In function 'get_u_int64_t':
/var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h:333:3: error: implicit declaration of function 'memcpy' [-Werror=implicit-function-declaration]
  333 |   memcpy(&tmp, X + O, sizeof(tmp));
      |   ^~~~~~
/var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h:330:1: note: include '<string.h>' or provide a declaration of 'memcpy'
  329 | #include <stdint.h>
  +++ |+#include <string.h>
  330 | static inline uint64_t get_u_int64_t(const uint8_t* X, int O)
/var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h:333:3: warning: incompatible implicit declaration of built-in function 'memcpy' [-Wbuiltin-declaration-mismatch]
  333 |   memcpy(&tmp, X + O, sizeof(tmp));
      |   ^~~~~~
/var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h:333:3: note: include '<string.h>' or provide a declaration of 'memcpy'
In file included from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_define.h:329,
                 from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/../../src/include/ndpi_main.h:34,
                 from /var/tmp/portage/net-firewall/nDPI-netfilter-20240320/work/nDPI-netfilter-20240320/ndpi-netfilter/src/ndpi_strcol.c:7:
/usr/lib/gcc/armv7a-unknown-linux-musleabihf/13/include/stdint.h:9:26: error: no include path in which to search for stdint.h
```

I trace this to /src/include/ndpi_define.h:329 :
```
#if defined(__arm__)
#include <stdint.h>
static inline uint64_t get_u_int64_t(const uint8_t* X, int O)
{
  uint64_t tmp;
  memcpy(&tmp, X + O, sizeof(tmp));
  return tmp;
}
#else
#define get_u_int64_t(X,O)  (*(u_int64_t *)((&(((u_int8_t *)X)[O]))))
#endif // __arm__

```

So we can't use stdint.h when compiling kernel modules. However, we can't comment this out either as we need get_u_int64_t(X,O) for compiling
src/lib/protocols/bgp.c:44

I'm not totally sure I understand what is going on here though? Is the issue simply that we need to avoid an endian conversion? 

Is there ever a world where sizeof(uint64_t) is not 4 bytes? Could we replace this with 4x u_int8_t assignments, ie our own basic memcpy?

**Comment by ewildgoose** [comment]
Additional. I think this minimal patch might correct the problem (I had overlooked that we can use kernel memcpy)

```
--- src/include/ndpi_define.h.orig	2024-04-04 13:40:26.743904063 +0000
+++ src/include/ndpi_define.h	2024-04-04 14:39:06.271096852 +0000
@@ -326,7 +326,7 @@
 #define get_u_int16_t(X,O)  (*(u_int16_t *)((&(((u_int8_t *)X)[O]))))
 #define get_u_int32_t(X,O)  (*(u_int32_t *)((&(((u_int8_t *)X)[O]))))
 #if defined(__arm__)
-#include <stdint.h>
+#include <linux/string.h>
 static inline uint64_t get_u_int64_t(const uint8_t* X, int O)
 {
   uint64_t tmp;
```

**Comment by vel21ripn** [comment]
IMHO sizeof(uint64_t) is always 8 bytes.
I don't understand the point of using memcpy(). The compiler knows about uint64_t and can do the copying itself.
If you remove the conditional compilation "if defined(\_\_arm\_\_)", will it build and work? Tests except two must be passed.

**Comment by ewildgoose** [comment]
Hi, I have the same basic question. I wondered if it was due to "endian" issues on arm? The use of this function is to make direct comparisons with a constant though, so I'm not even sure that it works as intended? 

The only thing I can assume is that it's directly copying the 4 bytes in big endian order and then it treats them as a little endian value on arm. I understand that the network order is fixed, but whether this is the correct solution is quite unclear to me? 

So I guess my patch doesn't change the "intent" of the original, it does the copy in byte order, but I don't have a test to show if this does what is wanted or not?

Edit: I might try and find the original commit upstream and see if that has more information?

**Comment by ewildgoose** [comment]
OK, so I think I understand.

Further down we have definitions for little endian reads. These are presumably intended to be big endian reads (although I have some doubts that it works correctly). 

The memcpy is because arm doesn't like non aligned memory accesses (on many architectures anyway). If you run git blame on the original file you can see various commits which explain in more detail.

In conclusion, I think the patch above is the correct solution here

**Comment by vel21ripn** [comment]
I would remove "#include <stdint.h>", "#include <string.h>" from ndpi_define.h 
and add them where required (src/include/ndpi_replace_printf.h, src/lib/third_party/src/ ndpi_sha256.c).
